src/goto-programs/remove_virtual_functions.cpp

```diff
@@ -144,11 +155,126 @@ static void create_static_function_call(
   }
 }
 
+/// If there are ASSERT or ASSUME instructions involving \p argument_for_this
+/// in the instructions directly preceding the virtual function call then we
+/// duplicate them for \p temp_var_for_this. This is mainly aimed at null
+/// checks, to make the local safe pointer analysis work. We stop as soon as we
+/// find an instruction that is not an ASSERT or ASSUME involving
+/// \p argument_for_this.
+/// \param goto_program: The goto program containing the virtual function call
+/// \param instr_it: Iterator to the virtual function call in \p goto_program
+/// \param argument_for_this: The original expression for the this argument of
+///   the virtual function call
+/// \param temp_var_for_this: The new expression for the this argument of the
+///   virtual function call
+/// \return A goto program consisting of all the amended asserts and assumes
+static goto_programt analyse_checks_directly_preceding_function_call(
+  const goto_programt &goto_program,
+  goto_programt::const_targett instr_it,
+  const exprt &argument_for_this,
+  const symbol_exprt &temp_var_for_this)
+{
+  goto_programt checks_directly_preceding_function_call;
+
+  while(instr_it != goto_program.instructions.cbegin())
+  {
+    instr_it = std::prev(instr_it);
+
+    if(!(instr_it->type == ASSERT || instr_it->type == ASSUME))
```

Comment:
Agree to skipping over assertions, but suggest only duplicating assumptions (duplicating assertions would be confusing to a user, as any violation would show up twice)